st2actions/st2actions/runners/localrunner.py

```diff
@@ -106,21 +109,40 @@ def run(self, action_parameters):
             args = 'chmod +x %s ; %s' % (script_local_path_abs, args)
 
         env = os.environ.copy()
+        # Make sure os.setsid is called on each spawned process so that all processes
+        # are in the same group.
         process = subprocess.Popen(args=args, stdin=None, stdout=subprocess.PIPE,
-                                   stderr=subprocess.PIPE, shell=True, env=env)
-
-        try:
-            exit_code = process.wait(timeout=self._timeout)
-        except subprocess.TimeoutExpired:
-            # Action has timed out, kill the process and propagate the error
-            # Note: process.kill() will set the returncode to -9 so we don't
-            # need to explicitly set it to some non-zero value
-            process.kill()
-            error = 'Action failed to complete in %s seconds' % (self._timeout)
-        else:
-            error = None
+                                   stderr=subprocess.PIPE, shell=True, env=env,
+                                   preexec_fn=os.setsid)
+
+        error_holder = {}
+
+        def on_timeout_expired(timeout):
+            try:
+                process.wait(timeout=self._timeout)
+            except subprocess.TimeoutExpired:
+                # Set the error prior to kill the process else the error is not picked up due
+                # to eventlet scheduling.
+                error_holder['error'] = 'Action failed to complete in %s seconds' % (self._timeout)
+                # Action has timed out, kill the process and propagate the error. The process
+                # is started as sudo -u {{system_user}} -- bash -c {{command}}. Introduction of the
+                # bash means that multiple independent processes are spawned without them being
+                # children of the process we have access to and this requires use of pkill.
+                # Ideally os.killpg should have done the trick but for some reason that failed.
+                # Note: pkill will set the returncode to 143 so we don't need to explicitly set
+                # it to some non-zero value.
+                try:
+                    killcommand = shlex.split('sudo pkill -TERM -s %s' % process.pid)
```

Comment:
Since we control the input here it might be a bit better and safer to just use a list here instead of `shlex.split` (`args = ['sudo', 'pkill', '-TERM", '-s', process.pid]`).

Using `shlex.split` means traditional shell injection should not be possible, but if a value of `process.pid` can be controlled, it can be made to pass multiple arguments to `pkill` which is not possible when using a list.